### PR TITLE
Add a filter handler after chooseHost phase.

### PR DESCRIPTION
**Pull request by wangfakang**
### Issues associated with this PR

In some scenarios, the filter needs to be executed after the `chooseHost`, such as modifying the request header based on the selected host.

### Solutions

Add a receiveFilter entry point `DownFilterAfterChooseHost`, which is executed after `ChooseHost ` and before `sendRequest`.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/proxy/downstream.go [hunk 1]
```diff
@@ -627,7 +656,8 @@ func (s *downStream) getUpstreamProtocol() (currentProtocol types.ProtocolName)
 	return currentProtocol
 }
 
-func (s *downStream) receiveHeaders(endStream bool) {
+func (s *downStream) chooseHost(endStream bool) {
```
Comment (nejisama) [inline pkg/proxy/downstream.go:659]: choose host 为什么需要endStream作为参数？

File: pkg/proxy/downstream.go [hunk 2]
```diff
@@ -1435,10 +1474,10 @@ func (s *downStream) processError(id uint32) (phase types.Phase, err error) {
 		return
 	}
 
-	if s.receiverFiltersAgain {
```
Comment (nejisama) [inline pkg/proxy/downstream.go:1438]: 这个逻辑修改是为什么

File: pkg/proxy/downstream.go (hunk 1 above)
Comment (wangfakang) [reply pkg/proxy/downstream.go:659]: > choose host 为什么需要endStream作为参数？

在direct response场景下，此时会跳过`receiveHeaders`、`receiveData`，需要提前设置`s.downstreamRecvDone = endStream` 用于`endStream()`中做buffer复用处理。

File: pkg/proxy/downstream.go (hunk 2 above)
Comment (wangfakang) [reply pkg/proxy/downstream.go:1438]: > 这个逻辑修改是为什么

receiverFiltersAgain类型从`bool`调整为`types.Phase`类型，使其作为一个通用的重试阶段变量（当前已有的重试：route、ChooseHost）。

**Pull request closed without merging**